lib/Relation.php

```diff
@@ -721,39 +811,51 @@ public function reverse_order(): array
      */
     private function firstOrLast(int $limit = null, bool $isAscending): array
     {
-        $this->limit($limit ?? 1);
+        $options = array_merge($this->options, ['limit' => $limit ?? 1]);
 
         $pk = $this->table()->pk;
         if (!empty($pk)) {
-            if (array_key_exists('order', $this->options)) {
+            if (array_key_exists('order', $options)) {
                 $reverseCommand = $isAscending ? 'DESC' : 'ASC';
 
-                if (str_contains($this->options['order'], implode(" {$reverseCommand}, ", $this->table()->pk) . " {$reverseCommand}")) {
-                    $this->options['order'] = SQLBuilder::reverse_order((string) $this->options['order']);
+                if (str_contains($options['order'], implode(" {$reverseCommand}, ", $this->table()->pk) . " {$reverseCommand}")) {
+                    $options['order'] = SQLBuilder::reverse_order((string) $options['order']);
                 }
-            } elseif (!array_key_exists('having', $this->options)) {
+            } elseif (!array_key_exists('having', $options)) {
                 $command = $isAscending ? 'ASC' : 'DESC';
-                $this->options['order'] = implode(" {$command}, ", $this->table()->pk) . " {$command}";
+                $options['order'] = implode(" {$command}, ", $this->table()->pk) . " {$command}";
             }
         }
 
-        return $this->to_a();
+        return $this->to_a_withOptions($options);
     }
 
     /**
-     * Converts relation objects to array.
+     * Converts relation objects to array with the currently set of options.
      *
      * @return array<TModel> All the rows that matches query. If no rows match, returns []
      */
     public function to_a(): array
+    {
+        return $this->to_a_withOptions($this->options);
+    }
+
+    /**
+     * Converts relation objects to array with a given options.
+     *
+     * @param RelationOptions $options
+     *
+     * @return array<TModel> All the rows that matches query. If no rows match, returns []
+     */
+    private function to_a_withOptions(array $options): array
```

Comment:
This is a bit awkward, isn't it? What are you fixing? If it's related to polluting the user's options with "order" in the case of "having", then you probably won't be surprised to hear that my advice is going to be to move that logic to `Table::options_to_sql`, and keep the rest of this clean.

Reply:
It's to:

1) Deny the user to be able to call `Author::to_a($myHandCraftedOptionsThatWouldHaveToBeInputChecked)`
2) Fix the bugs in find_by(), first(), last(), and take(). There are new tests in the test classes which trigger the bug of clobbering $options, but this submission fixes those bugs.

Reply:
> 1. Deny the user to be able to call `Author::to_a($myHandCraftedOptionsThatWouldHaveToBeInputChecked)`

Is that a thing? There's no static `to_a` on `Model` and the `to_a` on `Relation` doesn't take any arguments. But let me take a look at your tests...

Reply:
Ah, yep, those tests show the problem, for sure...

Reply:
Maybe a bit simpler?

https://github.com/php-activerecord/activerecord/compare/how-about-this?expand=1